### PR TITLE
GSDX: Improvements to the GSDX dialog

### PR DESCRIPTION
**Pull request by ssakash**
**Summary of changes:**

* Add new logos for the renderers. **( logos were made by [bmate](http://forums.pcsx2.net/User-bmate) )**

_**D3D9**_
D3D9

_**D3D11**_
D3D11

_**OpenGL**_
ogl

_**Null**_
null

* Gray out upscaling hacks on native resolution.

* Properly display usage of D3D10/D3D11 on the combobox. Also, Sort all the renderers in ascending order.

* Use highest available version of DX by default when GSDX dialog is used the first time (or) when GSDX.ini is not found.

* Remove all the extra NULL renderers with different rendering backends. ( There should be no difference between them and it makes maintaining renderer related code a bit easier)

**Build download link : [here](https://ci.appveyor.com/api/buildjobs/8m9je19co68kox92/artifacts/pcsx2-v1.5.0-dev-576-g51b413f-master-816-vs2015-Win32-AppVeyor.7z)**

### CODE REVIEW

**Comment by ssakash** [comment]
Initially I also wanted to add my ``integration of software renderers`` commit but didn't do it since it causes the _following conflicts_,

* dealing with F9 hotkey.
* In that case, we'd need new logo for Software renderer.

so.. yeah. it's better to handle it in the future with a bit of more thinking.

**Comment by refractionpcsx2** [comment]
I'm not a huge fan of the font used in the logos, it all looks a bit...skinny and undefined

**Comment by ssakash** [comment]
> I'm not a huge fan of the font used in the logos, it all looks a bit...skinny and undefined
GSDX 0 isn't so bad, its got a bit more bulk to it.

I'm not really much of a design expert but I think this makes all the logos consistent compared to the old style where the logos were different for all the renderers.

Also, I'm sure **bmate** will appreciate your constructive opinions by making the logos even better ;)

**Comment by refractionpcsx2** [comment]
I'm thinking something a bit more like this, so the chosen renderer is a bit bolder and looks a bit more like it means business (this is ofc a 2 min mockup in Photoshop :P )

gsdxogl

**Comment by willkuer** [comment]
If you change the renderer enum the gsdx.ini file will not be compatible.

**Comment by refractionpcsx2** [comment]
> If you change the renderer enum the gsdx.ini file will not be compatible.

won't it? It should be fine as all the code has been changed to reference the enum rather than using the number references.

**Comment by gregory38** [comment]
The ini file still contains an integer not an enum. So yeah it will broke everything.

**Comment by refractionpcsx2** [comment]
Won't it only be broken until you next save your settings?

**Comment by gregory38** [comment]
+ the fact that I'm not sure enum value have any guarantee to start at 0

**Comment by willkuer** [comment]
The enum value is stored as int/string and during ini-read casted to enum. I would recommend just to keep the manual enum assigment. It is a minor change and all the other improvements remain the same.

If you want to save the enum as enum one would probably save the datatype as well. This is at least how I do it in my project. I am using xmls instead of ini's though. Objects in ini's are awkward.

But as I said just revert the manal enum assignment and everything is good.

**Comment by willkuer** [comment]
I think you need to replace all renderer default by your best renderer to make it safe.

**Comment by ssakash** [comment]
> I think you need to replace all renderer default by your best renderer to make it safe.

I'm not sure what you mean by making it safe ? my intention was to only default to the best renderer when the dialog is first started without any INI file. also I don't think you can do the Direct3D level check on ``GS.h``

> The ini file still contains an integer not an enum. So yeah it will broke everything

If there are any unknown values present for ``Renderer`` then we'll fall back to the default renderer. by unknown I mean the values which aren't used at the Enum. After that, the there should be no problem. Is this method not ideal enough ?

> the fact that I'm not sure enum value have any guarantee to start at 0

After undefined is initialized to -1 , the succeeding elements of the enum will have a value which follows proper arithmetic progression with a difference of 1. 

So, let's assume I make an ``enum class GS : int8_t`` and initialize first element ``row`` as ``-2`` then the succeeding element will have a value of  -1.

**Comment by willkuer** [comment]
The question is what happens if somebody runs the emulator first. Then he would probably get a different renderer than when he starts the plugin configuration. Also what happens after he clicks apply or cancel, pcsx2 crashes and all unexpected stuff. I think its generally good always to have the same default value for a given configuration. But maybe this is just me.

**Comment by ssakash** [comment]
> The question is what happens if somebody runs the emulator first. Then he would probably get a different renderer than when he starts the plugin configuration. Also what happens after he clicks apply or cancel, pcsx2 crashes and all unexpected stuff. I think its generally good always to have the same default value for a given configuration. But maybe this is just me.

I actually checked my code for around one hour before making a pull request and didn't reproduce any issues when running the emulator first (or) when applying the settings at the dialog with unknown values at the renderer option on the INI. It fell back to the default renderer (D3D9) whenever any unknown value was used. ( that makes sense looking at the settings dialog and GS.cpp code)

Either way, I changed the Enum elements value similar to the older values. just to be on the safe side as willkuer said ;)

**Comment by ssakash** [comment]
Oh, forgot to mention. there's also a logo for D3D10, it'll only appear for users whose GPU supports DX10 with no DX11 features supported 

d3d10

So @refractionpcsx2 , don't forget to rework the D3D10 logo :P

**Comment by refractionpcsx2** [comment]
yeh ill do all the ones in the PSD :P

**Comment by refractionpcsx2** [comment]
Here we go, straight from the original PSD :)

gsdx - gsdx9
gsdx - gsdx10
gsdx - gsdx11
gsdx - gsdxnull
gsdx - gsdxogl

**Comment by willkuer** [comment]
Maybe OGL instead of OpenGL?

Also maybe DX9, DX10, DX11, OGL4.5?

**Comment by ssakash** [comment]
Updated the code. Thanks for the logos, bmate and refraction !

> Also maybe DX9, DX10, DX11, OGL4.5?

The ``DX`` of ``GSDX`` makes it redundant to mention DX once again. I prefer ``OpenGL`` over ``OGL``. why use a short form when the abbreviation fits in the logo ?

I don't think it'd be nice to include the 4.5 :P

**Comment by willkuer** [comment]
These were more questions. If the dx in gsdx points to directx and gsdx9 is the full qualified name what does gsdxOpengl mean? I just wanted to point out that gsdx could be a name while dx9...dx11,ogl4.5 are the full qualifiers of the renderer. But I don't really care. It's just slightly inconsistent.

Techally OpenGL does not fit into the logo. That's why its fontsize is shrinked. If you shrink the fontsize further you could get the whole oxford dictionary inside the space. OGL might fit with the same fontsize. But again it was just a remark and I am not really pushing for it. It's again just a matter of consistency.

**Comment by willkuer** [comment]
Does gsdx still work with any psx emulator? The last time I saw somebody tried it it failed (but on linux & epsxe IIRC) does pcsx work with the latest gsdx?

**Comment by gregory38** [comment]
Gsdx ps1 never worked on linux

**Comment by Blackbird88** [comment]
The config does work in ePSXe 2.0 but all games crash on startup
image

**Comment by ssakash** [comment]
@willkuer : Thanks for mentioning about GSDX with PS1 emulators ! It seems the new logos don't fit the dialog for PS1 mode. I'll fix it tommorow , going to sleep now. ( a bit past my bed time)

here

Yes, I can confirm. All games crash at startup as @Blackbird88 mentioned.

**Comment by FlatOutPS2** [comment]
v1.4.0. GSdx plugins work fine with PS1 emulators, but latest git versions crash epsxe.

**Comment by refractionpcsx2** [comment]
I think the reason it was OGL and not OpenGL was because it didn't fit in the PSX windows :P So for the PSX ones I've shortened it to OGL again, here we go, set to 262x71 which was the size @ssakash gave me

gsdxpsx - gsdx9
gsdxpsx - gsdx10
gsdxpsx - gsdx11
gsdxpsx - gsdxnull
gsdxpsx - gsdxogl

**Comment by willkuer** [comment]
But if it does not even work anymore.... can we maybe drop psx emulation? Maybe ping gabest and ask him?

**Comment by refractionpcsx2** [comment]
We did talk about splitting the PS1 and PS2 side aaaaaages ago, apparently that split never happened :p

**Comment by willkuer** [comment]
I think we are not talking about splitting... It's broken. We are talking about dropping it as it would maintain manpower to make it work again. But when you have the psd files you could adjust the subtitle later of you want and decide to drop.

**Comment by refractionpcsx2** [comment]
Yeah I have the PSD's.

turteli just pushed a bug fix for the PSX side, maybe it works now?

**Comment by turtleli** [comment]
Yeah, should be fixed now (IOW I forgot there was a psx interface).

Some questions:
 * do we have to distinguish between dx10 and dx11? We could just call it dx11 and avoid an additional logo that most users won't see.
 * perhaps I'm mistaken, but are we not removing DX9 so that it only exists in the legacy plugin (which still needs to be hooked up on Windows)?

**Comment by refractionpcsx2** [comment]
> do we have to distinguish between dx10 and dx11? We could just call it dx11 and avoid an additional logo that most users won't see.

Users who only have a DX10 card that isn't compatible with DX11 will see this logo, so making the DX10 one say DX11 would be very confusing to look at.

> perhaps I'm mistaken, but are we not removing DX9 so that it only exists in the legacy plugin (which still needs to be hooked up on Windows)?

DX9 exists in the legacy and current plugins, that isn't going away unless there is a reason to.  We did discuss removing it but concluded there wasn't much point unless it becomes a bane to maintain.

**Comment by FlatOutPS2** [comment]
> Yeah, should be fixed now (IOW I forgot there was a psx interface).
I can confirm it works again. :+1:

**Comment by ssakash** [comment]
I made some changes to the PSX dialog, logos size/placement should be fine now thanks to remaking the logos by @refractionpcsx2 

**D3D9**
h1

**D3D11**
h2

**Null**
h3

**Comment by turtleli** [comment]
I dunno if it's the screenshot or not, but the logos are distorted in the latest images.

> Remove SDL renderer from combobox since it was **deprecated** long ago.

I believe it's meant to say removed.


(And this is purely my opinion, but do we really need a logo for each renderer? Why not just 1 logo for everything, or even no logos?)

**Comment by refractionpcsx2** [comment]
> I dunno if it's the screenshot or not, but the logos are distorted/have artifacts. (EDIT: the actual logos seem ok)

what's probably happening is he gave me the wrong size and windows is resizing the image to fit, but of course windows doesn't do any sort of filtering, so it's probably just cutting chunks out.

**Comment by ssakash** [comment]
> what's probably happening is he gave me the wrong size and windows is resizing the image to fit

The size I mentioned was the same one as the older logo, so it should be right.

> (And this is purely my opinion, but do we really need a logo for each renderer? Why not just 1 logo for everything, or even no logos?)

I don't mind it either way, I just updated the logos since the style of the initial logos were inconsistent. but I think the logos make the dialog a bit livelier :P 

I'll ask gabest's opinion on the logos.

**Comment by refractionpcsx2** [comment]
@gabest11 you are being summoned  :p

**Comment by Trevonn** [comment]
Shouldn't you take this opportunity to change the name of the plug-in itself? It sounds a bit weird calling the OpenGL version "GSDX OpenGL".

**Comment by refractionpcsx2** [comment]
> Shouldn't you take this opportunity to change the name of the plug-in itself? It sounds a bit weird calling the OpenGL version "GSDX OpenGL".

It's always been called that, why can it not mean GS Deluxe? ;)

**Comment by DPersonalized** [comment]
> Shouldn't you take this opportunity to change the name of the plug-in itself? It sounds a bit weird calling the OpenGL version "GSDX OpenGL".

I think renaming the plugin would just lead to confusion, regardless of how good the new name might be.

**Comment by refractionpcsx2** [comment]
> I think renaming the plugin would just lead to confusion, regardless of how good the new name might be.

I dunno, while it might reduce confusion about if its DX only or not, we will just get more people going "Why is the new version of GSDX not included with PCSX2?" and considering the lack of "is GSDX directx only?" questions we currently have, it really isn't an issue :)

**Comment by turtleli** [comment]
> The size I mentioned was the same one as the older logo, so it should be right.

I believe the distorted logo in the psx emu configure dialog indicates something's not right.

File: plugins/GSdx/GSdx.rc [hunk 1]
```diff
@@ -175,7 +181,9 @@ STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSM
 CAPTION "Settings..."
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
-    CONTROL         IDB_LOGO9,IDC_LOGO9,"Static",SS_BITMAP,7,7,175,44
+    CONTROL         IDB_PSX_LOGO9,IDC_PSX_LOGO9,"Static",SS_BITMAP | SS_REALSIZECONTROL,7,7,175,42
```
Comment (willkuer) [inline plugins/GSdx/GSdx.rc]: Not sure if intended but you change the ... I guess height ... from 44 to 42... Maybe that is your scaling issue?

**Comment by refractionpcsx2** [comment]
It looks fine for me, had to convert them to BMP's as it doesn't like PNG's in VS. But they went in fine.

psxdialog

File: plugins/GSdx/GSdx.rc (hunk 1 above)
Comment (ssakash) [inline plugins/GSdx/GSdx.rc]: It was done to have the height constant for all the logos. It was weird to have only a single logo with a different height.

**Comment by ssakash** [comment]
> I believe the distorted logo in the psx emu configure dialog indicates something's not right.

Is the logo also distorted at the dialog when opened through any PSX Emulators (or) only at the screenshot ?

**Comment by refractionpcsx2** [comment]
> Is the logo also distorted at the dialog when opened through any PSX Emulators (or) only at the screenshot ?

mine was taken from inside epsxe, i didn't use the preview.

**Comment by turtleli** [comment]
> Is the logo also distorted at the dialog when opened through any PSX Emulators

Yes, I used the plugin from AppVeyor on epsxe. What willkuer pointed out is likely what's wrong.

**Comment by ssakash** [comment]
> Yes, I used the plugin from AppVeyor on epsxe. What willkuer pointed out is likely to be the issue.

I'm not sure whether that's the issue. Hmm, did you check whether height to 42 fixed the distortion ?

> mine was taken from inside epsxe, i didn't use the preview.

Could you check whether the issue is present with/without ``SS_REALSIZECONTROL`` on resource script file ? I think that's most likely the issue...

**Comment by refractionpcsx2** [comment]
> Could you check whether the issue is present with/without SS_REALSIZECONTROL on resource script file ? I think that's most likely the issue...

I can't right now, but I didn't use your code, I literally just replaced the existing pictures with the new ones. There is no filtering when a picture is resized which means there will be graphical glitches, so you changing the size is more than likely the problem.

**Comment by turtleli** [comment]
> Hmm, did you check whether height to 42 fixed the distortion ?

It's already at 42 in your branch. Some things to note:
 * SS_REALSIZECONTROL does rescale the image, but it's the easiest way to deal with different DPI settings.
 * Since dialog units are used to determine size and not pixels, you may suffer from graphical glitches anyway(depending on the image size), but it should not be as bad as your screenshots show.

**Comment by ssakash** [comment]
Removing ``SS_REALSIZECONTROL`` seems to resolve the issue. Another alternative would be to find the ideal dialog units (or) stretch the dialog and re-use the logos intended for main dialog. I prefer the first method since it saves me quite a bit of work :P 

> Since dialog units are used to determine size and not pixels, you may suffer from graphical glitches anyway(depending on the image size), but it should not be as bad as your screenshots show.

The difference wasn't really so obvious when I looked at my CRT monitor. (seriously)

**Comment by willkuer** [comment]
Could be interesting to check different DPI settings.

**Comment by gregory38** [comment]
What is the status?

**Comment by ssakash** [comment]
Turtleli: are you fine with the removal of ``SS_REALSIZECONTROL`` on PSX dialog ? ( I hope it doesn't cause any conflicts with different DPI settings... )

**Comment by turtleli** [comment]
Well, it won't scale at higher DPI settings, but it's technically not a regression.

I still find it silly to have 8 different logos (it provides 0 functionality and takes up some valuable space), but I'll assign this to ref ;)

**Comment by refractionpcsx2** [comment]
Thanks :P I guess if we can get it to chop the image rather than resize it, we could get away with just the 4 images, worth checking out i guess.  TBH though, it doesn't make a huge difference, it's not like the plugin is ridiculously massive

**Comment by turtleli** [comment]
Yeah, it's more of a personal objection and I assume most people disagree with my opinion, so I decided to push the responsibility onto you ;)

**Comment by refractionpcsx2** [comment]
> Yeah, it's more of a personal objection and I assume most people disagree with my opinion, so I decided to push the responsibility onto you ;)

aww, you're so sweet <3 :P

**Comment by gregory38** [comment]
So, is it ready to be merged ?

This way we can plan the future update :) Current stuff that windows might miss
* I already removed the depth option. And replace it by an hw hack => "Disable Depth Emulation"
* I added a new speed hack "Fast Texture Invalidation"

Then future planned change
* Remove the "Safe Accurate Blending" from the GUI.
* Move Preload Frame into main section and maybe rename "Preload Framebuffer"

**Comment by refractionpcsx2** [comment]
I guess it's fine for now, we can't do any changes yet even if we wanted to as Ssakash hasn't got a computer currently. So I guess if we want further modifications (like reducing the number of logos) we could do another PR later.

**Comment by refractionpcsx2** [comment]
Sorry, wrong button ;p

**Comment by gregory38** [comment]
Any ETA for the new computer? But yeah, it can be improved later.

**Comment by refractionpcsx2** [comment]
Best I can offer from our conversations is "this month" as he said "Next Month" on wednesday

**Comment by gregory38** [comment]
Ok. You're in charge of the merge ;)

**Comment by refractionpcsx2** [comment]
Done, we're doomed!

**Comment by ssakash** [comment]
About computer : not sure about the exact delivery date but I'll surely get it before 7th of May. I'm itching to code lol

**Pull request merged**